internetarchive/openlibrary
# Add Internet Archive cover option

Closes #9333.


Feature. Adds an option to use the book's cover from the Internet Archive site if applicable.

### Technical

Current progress:
✅ IA cover option is visible for books with an `ocaid` - i.e. `/books/OL24755423M/Odyssey_book_iv`
✅ IA cover option is not visible for books without an `ocaid` - i.e. `/books/OL9737752M/The_Odyssey`
✅ Selecting "Use this Image" successfully uploads the IA cover
✅ Entering a link or image file and selecting "Submit" successfully uploads the user-selected cover, not the IA one
✅ Error checks are modified to run only if the regular "Submit" button is clicked
✅ Button disabling is switched from built-in HTML to manual CSS to prevent IA upload failure
✅ Inline styles should be removed
✅ Hitting the "Enter" key should click the original "Submit" button, not the IA one
✅ Mysterious cropping in the event of an error message should be prevented
✅ Would be nice to do a little bit of CSS clean-up to distinguish the two buttons and improve form spacing
✅ Cover form divided into three parts and redesigned, unnecessary JavaScript validation removed
✅ Individual button loading indicators replaced with global loading indicator

### Testing

1. Log in (if not logged in)
2. Go to the cover modal for a book that has an `ocaid`, such as `/books/OL24755423M/Odyssey_book_iv`
3. You should see an option to select the Internet Archive cover
4. Selecting that option should upload the IA cover, even if a URL or file was included in the form
5. Selecting the original "Submit" button should upload the user-included URL or file, not the IA cover

### Screenshot

**Cover modal:**
<img width="624" alt="Cover modal with new option" src="https://github.com/internetarchive/openlibrary/assets/140550988/5733ec3b-5d02-415b-a85e-a2ab25ab906c">

**Work cover form (`/add-cover`):**
<img width="611" alt="Screenshot 2024-06-20 at 2 39 59 PM" src="https://github.com/internetarchive/openlibrary/assets/140550988/4417eb95-55cb-48da-a48e-dcb2c5193d43">


### Stakeholders

@cdrini
